### ISSUE TITLE
Performance: write benchmark.js tests for suspected slow code paths (i.e. see `#query-optimize`)

### ISSUE THREAD

**Issue by dmarcelino** [balderdashy/waterline#989]
Several performance issues have been noted in #343, #805, ~~#829~~, #880, #944, balderdashy/sails-postgresql#141, balderdashy/sails-mysql#161 and a possible factor aggravating these is the extensive use of `_.cloneDeep()` across waterline and its dependencies. Not counting adapters, waterline-criteria, anchor or waterline-schema waterline has [20 occurrences](https://github.com/balderdashy/waterline/search?utf8=✓&q=cloneDeep+path%3Alib&type=Code) of `_.cloneDeep()`.

We should investigate all of these and check which can be cached, which can be replaced with `_.clone()` and which can be avoided completely. If we can avoid some of these usages it's likely performance will improve.

We can use the [load tests](https://github.com/balderdashy/waterline-adapter-tests/tree/master/test/load) to benchmark changes and measure performance improvements.

**Comment by devinivy** [comment]
:+1: thanks for bringing this up.  Related, I've been interested in the various cloned/manipulated versions of the schema and when/how they're used: https://github.com/balderdashy/waterline/blob/master/lib/waterline.js#L107 using waterline-schema and https://github.com/balderdashy/waterline/blob/master/lib/waterline/core/index.js#L35 manipulated per-collection.

A lot of this cloning and computing could be solved, I think, but creating a more robust private API for common tasks like grabbing all associations on a collection, finding primary keys, etc.

**Comment by blah238** [comment]
:+1: I have also long suspected the (excessive?) use of cloning to be associated with the memory consumption and performance problems in Waterline, but not had time to investigate it in-depth.

**Comment by devinivy** [comment]
Someone want to create a flame graph of a particular use-case for us to pour over? :smile: 
http://blog.nodejs.org/2012/04/25/profiling-node-js/

**Comment by sailsbot** [comment]
Thanks for posting, @dmarcelino.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by mikermcneil** [comment]
@dmarcelino @devinivy @blah238 btw guys don't let ole sailsbot discourage you-- feel free to continue the discussion here.  I just want to make sure we keep all inactive issues closed so it's easier to monitor new stuff in our issue queue (i.e. to keep an eye out for bugs/security issues)

re: cloneDeep-- @particlebanana, @sgress454, and I did a bunch of profiling while we were working some kinks out of Treeline this spring and we confirmed that cloneDeep was indeed gobbling cycles.  In our case, it was because of userland code, but it is definitely something to keep an eye on.  Thanks for focusing the discussion into a single bucket and adding some great commentary-- let's use this as a jumping off point for PRs and any related discussion regarding cloneDeep optimizations in Waterline.

**Comment by nbiton** [comment]
I've made some additional attempts at optimization a few months ago at https://github.com/nbiton/waterline/tree/query-optimize
This time I did my best to avoid the breaking change I introduced with my last attempt.

The results were quite promising on Node v0.10.*, but for some obscure reason, it was much less significant after I upgraded to Node v0.12.*

This might help anyone who's interested to see which parts of the code I noticed to have a bigger role in the performance hit.

**Comment by mikermcneil** [comment]
@nbiton thanks for the help, and the repo!  @particlebanana has been doing a good amount of performance checking in adapters recently -- using benchmark.js for that.  As for the Waterline core features, I still think cloning is the big thing to watch out for, but before we even touch that, we'll need to put together some solid benchmarks that hit the suspected problem code paths.  Your work in query-optimize will be an excellent start on that front!

**Comment by nbiton** [comment]
:+1:
